### PR TITLE
Add arm64v8, ppc64le, s390x images of ubuntu:focal using aptman/qus

### PR DESCRIPTION
**Pull request by nulano**
For https://github.com/python-pillow/Pillow/issues/5028#issuecomment-721428190. Companion to https://github.com/python-pillow/Pillow/pull/5088.

Add arm64v8, ppc64le, s390x images of ubuntu:focal using aptman/qus.
Images will fail to push until https://github.com/python-pillow/Pillow/pull/5088 is merged due to new failures (libtiff was previously untested on big-endian, now exposes new failures).

These images are the same as `ubuntu-20.04-focal-amd64` but on different architectures and with `libimagequant-dev` added.
More information is in https://github.com/python-pillow/Pillow/pull/5088.

Building all images locally now requires running `docker run --rm --privileged aptman/qus -s -- -p` or similar first.

### CODE REVIEW

File: ubuntu-20.04-focal-arm64v8/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    && rm -rf /var/lib/apt/lists/*
+
+RUN useradd pillow && \
+    addgroup pillow sudo && \
+    mkdir /home/pillow && \
+    chown pillow:pillow /home/pillow
+
+RUN virtualenv -p /usr/bin/python3.8 --system-site-packages /vpy3 && \
+    /vpy3/bin/pip install --upgrade pip && \
+    /vpy3/bin/pip install cffi olefile pytest pytest-cov && \
```
Comment (hugovk) [inline ubuntu-20.04-focal-arm64v8/Dockerfile:40]: And let's also disable the pip cache to save a bit of extra space, like https://github.com/python-pillow/docker-images/pull/96.

**Review by hugovk — COMMENTED** [review]
Couple of small suggestions

**Comment by nulano** [comment]
Dockerfiles updated to match the #96 version.

**Comment by hugovk** [comment]
Thanks!

> Images will fail to push until [python-pillow/Pillow#5088](https://github.com/python-pillow/Pillow/pull/5088) is merged due to new failures (libtiff was previously untested on big-endian, now exposes new failures).

So this is failing until https://github.com/python-pillow/Pillow/pull/5088 is merged (needs test changes from there), but https://github.com/python-pillow/Pillow/pull/5088 is also failing until this is merged (needs images pushed from here). Is that right?

Can we split things up so we can merge and keep things green?

**Comment by hugovk** [comment]
From https://github.com/python-pillow/Pillow/pull/5088:

---

## Speed comparison (approximate)

| Architecture | Travis CI | GHA / Create Docker image | GHA / Test in Docker image | GHA Total |
|-|-|-|-|-|
| arm64    | 7 min           | 15 min 50 sec | 8 min 15 sec (1.18x Travis)   | 24 min 5 sec |
| ppc64le | 4 min 20 sec | 16 min (*)       | 9 min 15 sec (2.15x Travis)   | 25 min 15 sec (*) |
| s390x | 2 min 15 sec    | 14 min (*)        | 11 min 45 sec (5.22x Travis) | 25 min 45 sec (*) |

*: The ppc64le and s390x image builds currently fail a bit early due to new test failures, the install time will be slightly longer.

---

Thanks for the timings. They're not too bad, per repo, and definitely worth trying out. And this repo CI doesn't need to run as often as the main Pillow one.

We could use a technique we used on Travis to help speed up the builds as a whole. The [GHA docs](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) say:

> The order that you define a matrix matters. The first option you define will be the first job that runs in your workflow.

So we can kick off the slow ones first and make better use of idle parallel jobs for the quick ones, so we're not waiting around for the slow ones to finish.

```diff
diff --git a/.github/workflows/build.yml b/.github/workflows/build.yml
index 0bc5563..46ad119 100644
--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -12,6 +12,11 @@ jobs:
       matrix:
         target: [ "stable", "latest" ]
         image:
+          # Run slower jobs first to give them a headstart and reduce waiting time
+          - "ubuntu-20.04-focal-arm64v8"
+          - "ubuntu-20.04-focal-ppc64le"
+          - "ubuntu-20.04-focal-s390x"
+          # Then run the remainder
           - "alpine"
           - "amazon-2-amd64"
           - "arch"
@@ -22,9 +27,6 @@ jobs:
           - "fedora-33-amd64"
           - "ubuntu-18.04-bionic-amd64"
           - "ubuntu-20.04-focal-amd64"
-          - "ubuntu-20.04-focal-arm64v8"
-          - "ubuntu-20.04-focal-ppc64le"
-          - "ubuntu-20.04-focal-s390x"
```

**Comment by nulano** [comment]
> Thanks!
> 
> > Images will fail to push until [python-pillow/Pillow#5088](https://github.com/python-pillow/Pillow/pull/5088) is merged due to new failures (libtiff was previously untested on big-endian, now exposes new failures).
> 
> So this is failing until [python-pillow/Pillow#5088](https://github.com/python-pillow/Pillow/pull/5088) is merged (needs test changes from there), but [python-pillow/Pillow#5088](https://github.com/python-pillow/Pillow/pull/5088) is also failing until this is merged (needs images pushed from here). Is that right?

That's what I thought, but I just remembered there is also another problem. The "stable" jobs here will continue failing until the next release, and since the image push here is conditional on "stable" success, https://github.com/python-pillow/Pillow/pull/5088 won't succeed until then either. Would it make sense to change it to look at the "latest" run instead? That would also let the `.ci/install_raqm.sh` change take effect here earlier. Then the new failing stable jobs can be `exclude`d from the matrix until the next release.

> Can we split things up so we can merge and keep things green?

Sure, the main repo PR can be split into adding xfail markers first and adding the new jobs second.

**Comment by hugovk** [comment]
Or we could update the Pillow submodule here to point to current Pillow `master`.

A more cautious alternative would be to patch the test changes onto 8.0.1 (in a temporary branch), and update the submodule to point to that instead.

@radarhere What do you think?

**Comment by radarhere** [comment]
How about this - I've created a temporary branch here, https://github.com/python-pillow/docker-images/tree/qus (if I remember, I'll remove it sometime), to just push the new images to docker, leaving all the existing images unmodified. I've restarted https://github.com/python-pillow/Pillow/pull/5088, so that now passes apart from coverage.

From there, the simplest thing would seem to be to wait three weeks until 8.1.0 to merge this in.

Sound good?

**Comment by nulano** [comment]
That seems reasonable to me, but see my comment on the commit https://github.com/python-pillow/docker-images/commit/43e252e29aecc3c6a054da95495b4fa47acd61f8#r45093209

I would instead use `if: always()` for the push step.

**Comment by radarhere** [comment]
Ok, I've updated Pillow in the branch to use the version from https://github.com/python-pillow/Pillow/pull/5088 and restored tests. The docker images have been updated, and I've rerun the Pillow PR jobs.

**Comment by nulano** [comment]
Now that the stable builds have been updated to 8.1.0, there are no failures anymore.

File: ubuntu-20.04-focal-arm64v8/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+ADD depends /depends
+RUN cd /depends \
+    && ./install_imagequant.sh \
+    && ./install_raqm.sh
+
+USER pillow
+CMD ["depends/test.sh"]
+
+#docker run -v $TRAVIS_BUILD_DIR:/Pillow pythonpillow/ubuntu-20.04-focal-arm64v8
```
Comment (radarhere) [inline ubuntu-20.04-focal-arm64v8/Dockerfile:51]:
```suggestion
#docker run -v $GITHUB_WORKSPACE:/Pillow pythonpillow/ubuntu-20.04-focal-arm64v8
```
See https://github.com/python-pillow/docker-images/pull/106

File: ubuntu-20.04-focal-ppc64le/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+ADD depends /depends
+RUN cd /depends \
+    && ./install_imagequant.sh \
+    && ./install_raqm.sh
+
+USER pillow
+CMD ["depends/test.sh"]
+
+#docker run -v $TRAVIS_BUILD_DIR:/Pillow pythonpillow/ubuntu-20.04-focal-ppc64le
```
Comment (radarhere) [inline ubuntu-20.04-focal-ppc64le/Dockerfile:51]:
```suggestion
#docker run -v $GITHUB_WORKSPACE:/Pillow pythonpillow/ubuntu-20.04-focal-ppc64le
```

File: ubuntu-20.04-focal-s390x/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+ADD depends /depends
+RUN cd /depends \
+    && ./install_imagequant.sh \
+    && ./install_raqm.sh
+
+USER pillow
+CMD ["depends/test.sh"]
+
+#docker run -v $TRAVIS_BUILD_DIR:/Pillow pythonpillow/ubuntu-20.04-focal-s390x
```
Comment (radarhere) [inline ubuntu-20.04-focal-s390x/Dockerfile:51]:
```suggestion
#docker run -v $GITHUB_WORKSPACE:/Pillow pythonpillow/ubuntu-20.04-focal-s390x
```

**Comment by wiredfool** [comment]
Are we good here?

**Comment by hugovk** [comment]
We're good, thanks!

**Pull request merged**